include/pmacc/particles/algorithm/ForEach.hpp

```diff
@@ -22,157 +22,447 @@
-                            uint32_t const numPartcilesInSupercell = superCell.getNumParticles();
+        //! Policy to reverse iterate over particles/frames of a supercell.
+        struct Reverse
+        {
+            template<typename T_ParBox>
+            static DINLINE auto getFirst(T_ParBox& pb, DataSpace<T_ParBox::Dim> const& superCellIdx)
+            {
+                return pb.getLastFrame(superCellIdx);
+            }
 
+            template<typename T_ParBox>
+            static DINLINE auto next(T_ParBox& pb, typename T_ParBox::FramePtr const& framePtr)
+            {
+                return pb.getPreviousFrame(framePtr);
+            }
 
-                            // end kernel if we have no particles
-                            if(numPartcilesInSupercell == 0)
-                                return;
+            template<typename T_ForEach>
+            static DINLINE auto createCtx(T_ForEach& forEachParticleInFrame)
+            {
+                return lockstep::makeVar<lcellId_t>(forEachParticleInFrame, lcellId_t(0u));
+            }
+        };
 
-                            using FramePtr = typename T_ParBox::FramePtr;
-                            FramePtr frame = pb.getFirstFrame(superCellIdx);
+        /** Execute a particle or frame functor for each particle/frame.
+         *
+         * @tparam T_AccessTag Type of the tag to mark if the particle or frame interface will be used.
+         *                     valid options: detail::CallParticleFunctor or detail::CallFrameFunctor
+         * @tparam T_Order Iteration order. valid options: Forward or Reverse
+         * @tparam T_ParBox Type of the particle box to traverse.
+         * @tparam T_numWorkers Number of lockstep workers.
+         */
+        template<typename T_AccessTag, typename T_Order, typename T_ParBox, uint32_t T_numWorkers>
+        struct ForEachParticle
+        {
+        private:
+            static constexpr uint32_t dim = T_ParBox::Dim;
 
-                            // offset of the superCell (in cells, without any guards) to the origin of the local domain
-                            DataSpace<dim> const localSuperCellOffset = superCellIdx - mapper.getGuardingSuperCells();
+            using SuperCellSize = typename T_ParBox::FrameType::SuperCellSize;
+            static constexpr uint32_t frameSize = pmacc::math::CT::volume<SuperCellSize>::type::value;
+            static constexpr uint32_t numWorkers = T_numWorkers;
 
-                            auto accFunctor
-                                = functor(acc, localSuperCellOffset, lockstep::Worker<T_numWorkers>{workerIdx});
+            /** Number of frames to skip.
+             *
+             * If we have more worker than particles in a frame we skip frames depending on the worker id.
+             */
+            static constexpr uint32_t numFramesToSkip = (numWorkers + frameSize - 1u) / frameSize;
```

Comment:
I think this is confusing - does this variable actually mean the number of frames to skip + 1?  E.g. for `numWorkers == frameSize` its value will be 1, while we should skip 0 frames in this case.

(Same for the twin of this variable inside the lambda)

Reply:
You are right the distance to the next frame. Letzt check this tomorrow offline and find a better naming.

Reply:
I will change it to `frameDistance`